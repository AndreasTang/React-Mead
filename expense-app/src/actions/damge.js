import uuid from 'uuid'

const addDamge = ({ description = '', note = '', amount = '', createdAt = 0 } = {}) => ({
    type: 'ADD_DAMGE',
    damge: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
})

const removeDamge = ({ id } = {}) => ({
    type: 'REMOVE_DAMGE',
    id
})

const editDamge = (id, editDamge ) => ({
    type: 'EDIT_DAMGEAMOUNT',
    id,
    editDamge
})


export { addDamge, removeDamge, editDamge }