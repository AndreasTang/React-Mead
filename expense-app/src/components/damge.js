import React from 'react'
import DamgeList from './damgeList'
import DamgeListFilter from './damgeListFilter'

const Damge = () => (
    <div>
        <DamgeListFilter/>
        <DamgeList/>
    </div>
)

export default Damge