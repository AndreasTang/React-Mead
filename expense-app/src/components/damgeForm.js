import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/initialize';

class DamgeForm extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
        description: props.damge ? props.damge.description : '',
        note: props.damge ? props.damge.note : '',
        amount: props.damge ? (props.damge.amount / 100).toString() : '',
        createdAt: props.damge ? moment(props.damge.createdAt) : moment(),
        datePickerFocused: false,
        error: false
        }
    }
    textHandler = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    noteHandler = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    amountHandler = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ datePickerFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault()
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: true }))
        } else {
            this.setState(() => ({ error: false }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                note: this.state.note,
                createdAt: this.state.createdAt.valueOf()
            })
        }
    }
    render() {
        return (
            <div>
                {this.state.error ? <p>Please provide description and amount</p> : <p></p>}
                <form onSubmit = {this.onSubmit}>
                    <p><input type = "text"
                    placeholder = "description"
                    autoFocus
                    value = {this.state.description}
                    onChange = {this.textHandler}
                    /></p>
                    <p><input
                    type = "number"
                    placeholder = "amount"
                    value = {this.state.amount}
                    onChange = {this.amountHandler}
                    /></p>
                    <SingleDatePicker
                    date = {this.state.createdAt}
                    onDateChange = {this.onDateChange}
                    focused = {this.state.datePickerFocused}
                    onFocusChange = {this.onFocusChange}
                    id = "CalendarSingleDatePicker"
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                    />
                    <p><textarea
                    placeholder = "note"
                    value = {this.state.note}
                    onChange = {this.noteHandler}
                    /></p>
                    <button>Submit</button>
                </form>
            </div>
        )
    }
}

export default DamgeForm