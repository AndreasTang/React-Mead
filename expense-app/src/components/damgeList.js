import React from 'react'
import { connect } from 'react-redux'
import DamgeListItem from './damgeListItem'
import getVisibleDamge from '../selectors/damge'

const DamgeList = (props) => (
    <div>
        <h1>Damge List</h1>
        {props.damge.length === 0 ? (
            <p>No Damge</p>
        ) : (
            props.damge.map((damgeItem) => {
                return <DamgeListItem key = {damgeItem.id} {...damgeItem} />
            })
        )}
        
    </div>
)

const mapStateToProps = (state) => {
    return {
        damge: getVisibleDamge(state.damge, state.filter)
    }
}

export { DamgeList }
export default connect(mapStateToProps)(DamgeList)