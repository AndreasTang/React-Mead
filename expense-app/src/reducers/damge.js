const damgeReducerDefaultState = []

const damgeReducer = (state = damgeReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_DAMGE': {
            return [...state, action.damge]
        }
        case 'REMOVE_DAMGE': {
            return state.filter(({ id }) => id !== action.id)
        }
        case 'EDIT_DAMGEAMOUNT' : {
            return state.map((damge) => {
                if (damge.id === action.id) {
                    return {
                        ...damge,
                        ...action.editDamge
                    }
                } else {
                    return damge
                }
            })
        }
        default: 
            return state
    }
}

export default damgeReducer