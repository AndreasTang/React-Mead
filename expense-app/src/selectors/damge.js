import moment from 'moment'

const getVisibleDamge = (damge, { text, sortBy, startDate, endDate }) => {
    return damge.filter((damge) => {
        const startDateMatch = startDate ? startDate.isSameOrBefore(moment(damge.createdAt), 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(moment(damge.createdAt), 'day') : true
        const textMatch = damge.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt > b.createdAt ? -1 : 1
        } else if (sortBy === 'amount') {
            return a.amount > b.amount ? -1 : 1
        }
    })
}

export default getVisibleDamge