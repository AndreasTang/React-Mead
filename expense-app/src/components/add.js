import React from 'react'
import DamgeForm from './damgeForm'
import { connect } from 'react-redux'
import { addDamge } from '../actions/damge'

export class Add extends React.Component {
    onSubmit = (damge) => {
        this.props.add(damge)
        this.props.history.push('./')
    }
    render() {
        return (
            <div>
                <DamgeForm onSubmit = {this.onSubmit} />
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        add: (damge) => {
            return dispatch(addDamge(damge))
        }
    }
}

export default connect(undefined, mapDispatchToProps)(Add)