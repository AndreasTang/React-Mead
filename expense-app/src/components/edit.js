import React from 'react'
import DamgeForm from './damgeForm'
import { connect } from 'react-redux'
import { editDamge, removeDamge } from '../actions/damge'

export class Edit extends React.Component {

    onSubmit = (damge) => {
        this.props.editDamge(this.props.damge.id, damge)
        this.props.history.push('/')
    }

    onClick = () => {
        this.props.removeDamge({ id: this.props.damge.id })
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <DamgeForm 
                damge = {this.props.damge}
                onSubmit = {this.onSubmit}
                />
                <button onClick = {this.onClick}
                >Remove</button>
            </div>
        )
    }
}

const mapStateToProps = (state, props) => {
    return {
        damge: state.damge.find((damge) => {
            return damge.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        editDamge: (id, damge) => {
            return dispatch(editDamge(id, damge))
        },
        removeDamge: (data) => {
            return dispatch(removeDamge(data))
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Edit)