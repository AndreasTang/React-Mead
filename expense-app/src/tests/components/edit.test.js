import React from 'react'
import { shallow } from 'enzyme'
import { Edit } from '../../components/edit'
import damge from '../fixtures/damge'

let edit, remove, history, wrapper

beforeEach(() => {
    edit = jest.fn()
    remove = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(<Edit editDamge = {edit} removeDamge = {remove} history = {history} damge = {damge[0]} />)
})

test('Testing should Edit page render correctly', () => {
    expect(wrapper).toMatchSnapshot()
})

test('Testing should handle editDamge', () => {
    wrapper.find('DamgeForm').prop('onSubmit')(damge[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(edit).toHaveBeenLastCalledWith(damge[0].id, damge[0])
})

test('Testing should handle removeDamge', () => {
    wrapper.find('button').prop('onClick')(damge[0])
    expect(history.push).toHaveBeenLastCalledWith('/')
    expect(remove).toHaveBeenLastCalledWith({id: damge[0].id})
})