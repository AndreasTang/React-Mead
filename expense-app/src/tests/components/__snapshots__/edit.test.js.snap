// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing should Edit page render correctly 1`] = `
<div>
  <DamgeForm
    damge={
      Object {
        "amount": 10,
        "createdAt": -432000000,
        "description": "kicks",
        "id": "1",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
